lib/Partitioner/Partitioner.cpp

```diff
@@ -139,6 +139,147 @@ static BFSLevel getBFSLevel(Function *F) {
   return bfs;
 }
 
+// Combine the partitions if necessary : if all outside uses of the nodes in
+// partition1 is in partition2, and the sum of memory comsuption of partition1
+// and partition2 is less than availableMemory, combine partition1 and
+// partition2.
+void Partitioner::partitionsCombine(NodeToFunctionMap &partitions,
+                                    FunctionToNodesMap &nodesSet,
+                                    uint64_t availableMemory) {
+
+  for (PartitionNodesMap::iterator it = nodesSet.begin(); it != nodesSet.end();
+       ++it) {
+    std::vector<Node *> outUsers = getOutUsers((*it).second);
+    if (outUsers.empty()) {
+      continue;
+    }
+
+    bool flag = true;
+    for (int i = 1, e = outUsers.size(); i < e; i++) {
+      if (partitions[outUsers[i]] != partitions[outUsers[i - 1]]) {
+        flag = false;
+        break;
+      }
+    }
+    if (flag) {
+      // This partition only has one successor.
+      Function *cur = (*it).first;
+      Function *suc = partitions[outUsers[0]];
+      NodesSet tmp = nodesSet.lookup(suc);
+      GraphMemInfo cost1 = partitions.getCost(cur);
+      GraphMemInfo cost2 = partitions.getCost(suc);
+      if (cost1.constMem + cost1.inMem + cost2.constMem + cost2.inMem -
+              cost1.outMem <
+          availableMemory) {
+        // We can combine the two partitions to fit one device.
+        for (NodesSet::iterator it2 = tmp.begin(); it2 != tmp.end(); ++it2) {
+          partitions.add(*it2, cur);
+        }
+        (*it).second.insert(tmp.begin(), tmp.end());
+        partitions.deletePartition(suc);
```

Comment:
The logic for partition removals could be moved into its own helper method. It could be useful at other places as well, I guess.